libmogwai-schedule/clock-system.c

```diff
@@ -0,0 +1,190 @@
+                                                  GSourceFunc     alarm_func,
+                                                  gpointer        user_data,
+                                                  GDestroyNotify  destroy_func);
+static void       mws_clock_system_remove_alarm  (MwsClock       *clock,
+                                                  guint           id);
+
+/**
+ * MwsClockSystem:
+ *
+ * Implementation of #MwsClock which uses the system wall clock.
+ *
+ * Since: 0.1.0
+ */
+struct _MwsClockSystem
+{
+  GObject parent;
+
+  GMainContext *context;  /* (owned) */
+  GPtrArray *alarms;  /* (owned) (element-type GSource) */
+};
+
+G_DEFINE_TYPE_WITH_CODE (MwsClockSystem, mws_clock_system, G_TYPE_OBJECT,
+                         G_IMPLEMENT_INTERFACE (MWS_TYPE_CLOCK,
+                                                mws_clock_system_clock_init))
+static void
+mws_clock_system_class_init (MwsClockSystemClass *klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->finalize = mws_clock_system_finalize;
+}
+
+static void
+mws_clock_system_clock_init (MwsClockInterface *iface)
+{
+  iface->get_now_local = mws_clock_system_get_now_local;
+  iface->add_alarm = mws_clock_system_add_alarm;
+  iface->remove_alarm = mws_clock_system_remove_alarm;
+}
+
+static void
+destroy_and_unref_source (GSource *source)
+{
+  g_source_destroy (source);
+  g_source_unref (source);
+}
+
+static void
+mws_clock_system_init (MwsClockSystem *self)
+{
+  self->context = g_main_context_ref_thread_default ();
+  self->alarms = g_ptr_array_new_with_free_func ((GDestroyNotify) destroy_and_unref_source);
+}
+
+static void
+mws_clock_system_finalize (GObject *obj)
+{
+  MwsClockSystem *self = MWS_CLOCK_SYSTEM (obj);
+
+  g_clear_pointer (&self->alarms, g_ptr_array_unref);
+  g_clear_pointer (&self->context, g_main_context_unref);
+
+  G_OBJECT_CLASS (mws_clock_system_parent_class)->finalize (obj);
+}
+
+static GDateTime *
+mws_clock_system_get_now_local (MwsClock *clock)
+{
+  return g_date_time_new_now_local ();
+}
+
+static guint
+mws_clock_system_add_alarm (MwsClock       *clock,
+                            GDateTime      *alarm_time,
+                            GSourceFunc     alarm_func,
+                            gpointer        user_data,
+                            GDestroyNotify  destroy_func)
+{
+  MwsClockSystem *self = MWS_CLOCK_SYSTEM (clock);
+
+  g_autoptr(GDateTime) now = mws_clock_system_get_now_local (clock);
+
+  /* If the @alarm_time is in the past, invoke the callback on the next main
+   * context iteration. */
+  GTimeSpan interval = g_date_time_difference (alarm_time, now);
+  if (interval < 0)
+    interval = 0;
+
+  g_autoptr(GSource) source = g_timeout_source_new_seconds (interval / G_USEC_PER_SEC);
+  g_source_set_callback (source, alarm_func, user_data, destroy_func);
+  g_source_attach (source, self->context);
+
+  guint id = GPOINTER_TO_UINT (source);
```

Comment:
Good catch. I’d neglected `g_source_get_id()` before because the [documentation was not convincing](https://gitlab.gnome.org/GNOME/glib/merge_requests/128), but it should do the trick.

Reply:
Fixup pushed. Let’s see what Jenkins says.